language/move-prover/tests/sources/stdlib/modules/libra_account.move

```diff
@@ -99,10 +99,10 @@ module LibraAccount {
     }
     spec fun deposit {
         aborts_if to_deposit.value == 0;
-        aborts_if !exists<T>(sender());
+        aborts_if !::exists<T>(sender());
```

Comment:
It would be good to check with the Prover team (@wrwg) to make sure this aligns with their expectations.

Reply:
Thanks Bob for pointing this out. I missed this one.

`exists` is one of our builtin functions, like your `move_to_sender` on the programming level. We choose to don't model builtins as special cases in the frontend. We need to maintain at least for such builtin functions that they can be used without leading `::`.  Also for spec helper functions, we would prefer not to see a `::`, but we would have more flexibility there. Do you have a discussion somewhere why you require this? I know there are reasons why someone could be concerned about adoption a Rust or C++ namespace model, I'm not sure they justify such a killer on usability and modularity.

Reply:
Technical note: if the restriction is on syntax level, perhaps what we need to have is a `BuiltinCall` variant. This would allow unqualified arbitrary function calls, but the call is resolved always against well-known targets in the code, not user functions.

Reply:
> We need to maintain at least for such builtin functions that they can be used without leading ::.

This isn't how Move works currently in that built-in functions are not keyworded. The issue here is that `LibraAccount` defines its own function called `exists` so `exists(...)` with no proceeding `::` refers to `LibraAccount::exists`, to then refer to the builtin you need the preceding `::` to indicate it is the one defined in the top level name space, not in the current module. 

The alternative would be to do something like 
```
spec fun deposit {
    use ::exists;
    aborts_if to_deposit.value == 0;
    aborts_if !exists<T>(sender());
    ...
}
```
But I do not have aliasing for top-level namespace items implemented yet. (but we should probably consider adding it)


>  if the restriction is on syntax level, perhaps what we need to have is a BuiltinCall variant. This would allow unqualified arbitrary function calls, but the call is resolved always against well-known targets in the code, not user functions.

I'm not sure if I follow the suggestion exactly, but this is exactly what `::exists` is doing, it forces the call into being one of in the top-level namespace (which currently only contains builtins)

Reply:
> This isn't how Move works currently in that built-in functions are not keyworded.

This has nothing to do with Move but with your implementation. In our implementation, there is no difference for the type checker and the AST between builtin and user defined functions. For the user there still is. This technique is quite common in compilers, I have used it for multiple languages, and it reduces a lot of boilerplate in my experience.

`+` is also just a regular function for the type checker and our internal AST.

This PR is currently unacceptable for the Move prover team, so I would ask you to not submit until we discussed it (I wished to have known of this fundamental change to specs earlier.)

Reply:
> The issue here is that LibraAccount defines its own function called exists so exists(...) with no proceeding :: refers to LibraAccount::exists

So does this mean the :: is only needed because LibraAccount::exists is defined in the Move program? If that is the case, the issue would be less severe than I thought. In this case the real issue is that you resolving spec expressions in the same name scope like regular expressions. But those scopes are completely disjoint. A spec expression can never call a Move function, and a move expression can never call a spec function.

Reply:
> So does this mean the :: is only needed because LibraAccount::exists is defined in the Move program?

That is correct. It is only needed because it is defined in the current module

>  A spec expression can never call a Move function

But it can refer to them, no? Or can it only refer to the function/struct it is currently spec-ing?

Reply:
One can only refer to move functions from the same module in `spec fun some_func { ... }`, but not anywhere in expressions. Structs are different because of type parameters.